platform/viewer/src/components/ViewportGrid.tsx

```diff
@@ -22,136 +23,106 @@ const ORIENTATION_MAP = {
   },
 };
 
+const createHpInfo = (protocol, stage, activeStudyUID) => {
+  return {
+    hangingProtocolId: protocol.id,
+    stageId: stage.stageId,
+    stageIdx: protocol.stages.findIndex(it => it === stage),
+    activeStudyUID,
+  };
+};
+
+const compareViewportOptions = (opts1, opts2) => {
+  if ((opts1.viewportType || 'stack') != opts2.viewportType) {
+    return false;
+  }
+  return true;
+};
+
 function ViewerViewportGrid(props) {
   const { servicesManager, viewportComponents, dataSource } = props;
   const [viewportGrid, viewportGridService] = useViewportGrid();
 
-  const { numCols, numRows, activeViewportIndex, viewports } = viewportGrid;
+  const { layout, activeViewportIndex, viewports } = viewportGrid;
+  const { numCols, numRows } = layout;
 
   // TODO -> Need some way of selecting which displaySets hit the viewports.
   const {
     displaySetService,
     measurementService,
     hangingProtocolService,
     uiNotificationService,
-  } = servicesManager.services;
+  } = (servicesManager as ServicesManager).services;
 
   /**
-   * This callback runs only after displaySets have changed (created and added or modified)
-   */
-  const updateDisplaySetsForViewports = useCallback(
-    availableDisplaySets => {
-      if (!availableDisplaySets.length) {
-        return;
-      }
+   * This callback runs after the viewports structure has changed in any way.
+   * On initial display, that means if it has changed by applying a HangingProtocol,
+   * while subsequently it may mean by changing the stage or by manually adjusting
+   * the layout.
 
-      const {
-        viewportMatchDetails,
-        hpAlreadyApplied,
-      } = hangingProtocolService.getMatchDetails();
+   */
+  const updateDisplaySetsFromProtocol = (
+    protocol,
+    stage,
+    activeStudyUID,
+    viewportMatchDetails
+  ) => {
+    const availableDisplaySets = displaySetService.getActiveDisplaySets();
+
+    if (!availableDisplaySets.length) {
+      console.log('No available display sets', availableDisplaySets);
+      return;
+    }
 
-      if (!viewportMatchDetails.size) {
+    // Match each viewport individually
+    const { layoutType } = stage.viewportStructure;
+    const stageProps = stage.viewportStructure.properties;
+    const { columns: numCols, rows: numRows, layoutOptions = [] } = stageProps;
+
+    /**
+     * This find or create viewport uses the hanging protocol results to
+     * specify the viewport match details, which specifies the size and
+     * setup of the various viewports.
+     */
+    const findOrCreateViewport = viewportIndex => {
```

Comment:
There are two implementations of `findOrCreateViewport` - this one here and one in `findViewportByPosition.ts`. Do we need both?

Reply:
Ok.